desc/integrals/basis.py

```diff
@@ -58,6 +85,627 @@ def _in_epigraph_and(is_intersect, df_dy_sign, /):
+        cheb : jnp.ndarray
+            Shape (..., M, N).
+            Chebyshev coefficients to use. If not given, uses ``self.cheb``.
+
+        Returns
+        -------
+        f : jnp.ndarray
+            Shape z.shape.
+            Chebyshev series evaluated at z.
+
+        """
+        cheb = _chebcast(setdefault(cheb, self.cheb), z)
+        N = cheb.shape[-1]
+        x_idx, y = self.isomorphism_to_C2(z)
+        y = bijection_to_disc(y, self.domain[0], self.domain[1])
+        # Chebyshev coefficients αₙ for f(z) = ∑ₙ₌₀ᴺ⁻¹ αₙ(x[z]) Tₙ(y[z])
+        # are held in cheb with shape (..., num cheb series, N).
+        cheb = jnp.take_along_axis(cheb, x_idx[..., jnp.newaxis], axis=-2)
+        f = idct_non_uniform(y, cheb, N)
+        assert f.shape == z.shape
+        return f
+
+    def intersect2d(self, k=0.0, *, eps=_eps):
+        """Coordinates yᵢ such that f(x, yᵢ) = k(x).
+
+        Parameters
+        ----------
+        k : jnp.ndarray
+            Shape must broadcast with (..., *cheb.shape[:-1]).
+            Specify to find solutions yᵢ to f(x, yᵢ) = k(x). Default 0.
+        eps : float
+            Absolute tolerance with which to consider value as zero.
+
+        Returns
+        -------
+        y : jnp.ndarray
+            Shape (..., *cheb.shape[:-1], N - 1).
+            Solutions yᵢ of f(x, yᵢ) = k(x), in ascending order.
+        is_intersect : jnp.ndarray
+            Shape y.shape.
+            Boolean array into ``y`` indicating whether element is an intersect.
+        df_dy_sign : jnp.ndarray
+            Shape y.shape.
+            Sign of ∂f/∂y (x, yᵢ).
+
+        """
+        c = _subtract_first(_chebcast(self.cheb, k), k)
+        # roots yᵢ of f(x, y) = ∑ₙ₌₀ᴺ⁻¹ αₙ(x) Tₙ(y) - k(x)
+        y = chebroots_vec(c)
+        assert y.shape == (*c.shape[:-1], self.N - 1)
+
+        # Intersects must satisfy y ∈ [-1, 1].
+        # Pick sentinel such that only distinct roots are considered intersects.
+        y = _filter_distinct(y, sentinel=-2.0, eps=eps)
+        is_intersect = (jnp.abs(y.imag) <= eps) & (jnp.abs(y.real) < 1.0)
+        # Ensure y ∈ (-1, 1), i.e. where arccos is differentiable.
+        y = jnp.where(is_intersect, y.real, 0.0)
+
+        # TODO: Multipoint evaluation with FFT.
+        #   Chapter 10, https://doi.org/10.1017/CBO9781139856065.
+        n = jnp.arange(self.N)
+        #      ∂f/∂y =      ∑ₙ₌₀ᴺ⁻¹ aₙ(x) n Uₙ₋₁(y)
+        # sign ∂f/∂y = sign ∑ₙ₌₀ᴺ⁻¹ aₙ(x) n sin(n arcos y)
+        df_dy_sign = jnp.sign(
+            jnp.linalg.vecdot(
+                n * jnp.sin(n * jnp.arccos(y)[..., jnp.newaxis]),
+                self.cheb[..., jnp.newaxis, :],
+            )
+        )
+        y = bijection_from_disc(y, self.domain[0], self.domain[-1])
+        return y, is_intersect, df_dy_sign
+
+    def intersect1d(self, k=0.0, *, num_intersect=None, pad_value=0.0):
+        """Coordinates z(x, yᵢ) such that fₓ(yᵢ) = k for every x.
```

Comment:
some sort of global comment would be useful explaining how x, y, z, f etc map to rho, alpha, zeta, B etc.

Reply:
done after next commit